Build failure: python3 cross

`python3: 3.11.9 -> 3.12.4` (737b5cd4f2f0113727312a12b4dfceee877a88c3) regressed cross compilation for python3. 

### Steps To Reproduce

Steps to reproduce the behavior:
1. `nix-build -A pkgsCross.aarch64-multiplatform.python3`

### Build log

```
error: builder for '/nix/store/01a9nj8xa7ly11amy2riigp7p318pm4m-python3-aarch64-unknown-linux-gnu-3.12.4.drv' failed with exit code 1;
       last 10 log lines:
       > In file included from ./Include/internal/pycore_object.h:13,
       >                  from Objects/cellobject.c:4:
       > ./Include/internal/pycore_interp.h:193:24: error: field 'dtoa' has incomplete type
       >   193 |     struct _dtoa_state dtoa;
       >       |                        ^~~~
       > make: *** [Makefile:2737: Objects/abstract.o] Error 1
       > make: *** [Makefile:2737: Parser/tokenizer.o] Error 1
       > make: *** [Makefile:2737: Objects/bytesobject.o] Error 1
       > make: *** [Makefile:2737: Objects/cellobject.o] Error 1
       > make: *** [Makefile:2737: Objects/bytearrayobject.o] Error 1
       For full logs, run 'nix log /nix/store/01a9nj8xa7ly11amy2riigp7p318pm4m-python3-aarch64-unknown-linux-gnu-3.12.4.drv'.
```

### Additional context

```
737b5cd4f2f0113727312a12b4dfceee877a88c3 is the first bad commit
commit 737b5cd4f2f0113727312a12b4dfceee877a88c3
Author: Martin Weinelt <hexa@darmstadt.ccc.de>
Date:   Wed Jun 19 04:03:40 2024 +0200

    python3: 3.11.9 -> 3.12.4
    
    As per our documented update schedule, we'll default to Python 3.12 in
    the 24.11 release cycle, while recursing into both 3.11 and 3.12.

 pkgs/development/interpreters/python/default.nix | 30 +++++++++++++++---------------
 pkgs/top-level/all-packages.nix                  |  4 ++--
 2 files changed, 17 insertions(+), 17 deletions(-)
❯ git bisect log
git bisect start
# status: waiting for both good and bad commits
# good: [cbe5cf5d81dee7911aa8f88f143889f2abeae871] buildRustCrate: don't swallow build.rs output
git bisect good cbe5cf5d81dee7911aa8f88f143889f2abeae871
# status: waiting for bad commit, 1 good commit known
# bad: [79316737d60c607c023eb5a68526d9a560229fd4] Merge pull request #325011 from r-ryantm/auto-update/dkimpy
git bisect bad 79316737d60c607c023eb5a68526d9a560229fd4
# bad: [b9ebc70b93c7bcc81000d2214768079225e9224a] python312Packages.python-kasa: 0.6.2.1 -> 0.7.0.2
git bisect bad b9ebc70b93c7bcc81000d2214768079225e9224a
# bad: [43c7c50717602d54f82ecc579ade0287241a4b16] python3Packages.deepdiff: 6.7.1 -> 7.0.1
git bisect bad 43c7c50717602d54f82ecc579ade0287241a4b16
# good: [f822b2ba5c97189cd5cfcf6364dbf5318cfcba27] Merge remote-tracking branch 'origin/staging-next' into staging
git bisect good f822b2ba5c97189cd5cfcf6364dbf5318cfcba27
# good: [ae3742c661e9f6067f11295fdc09714494777c58] Merge pull request #315740 from mweinelt/xz-5.6.2
git bisect good ae3742c661e9f6067f11295fdc09714494777c58
# bad: [520dca3b6eefe88a02f0af4752b809399c7bb2f8] python3Packages.awslambdaric: 2.0.10 -> 2.0.11
git bisect bad 520dca3b6eefe88a02f0af4752b809399c7bb2f8
# bad: [e4676179b3325ef268d63789d60522acee78428d] python311Packages.pluggy: 1.4.0 -> 1.5.0
git bisect bad e4676179b3325ef268d63789d60522acee78428d
# good: [12df9ba6c373c7b864683904adc585acf98dddaa] Merge #321340: gcc12: 12.3.0 -> 12.4.0
git bisect good 12df9ba6c373c7b864683904adc585acf98dddaa
# good: [1abd769c912f7d7196d662f80887f9e1bcf0f1d5] Merge #320434: nettle: 3.9.1 -> 3.10
git bisect good 1abd769c912f7d7196d662f80887f9e1bcf0f1d5
# bad: [737b5cd4f2f0113727312a12b4dfceee877a88c3] python3: 3.11.9 -> 3.12.4
git bisect bad 737b5cd4f2f0113727312a12b4dfceee877a88c3
# good: [2ae2c768200fc23b00565be98d2001e1cc9ef7ff] Merge #316547: findutils: 4.9.0 -> 4.10.0
git bisect good 2ae2c768200fc23b00565be98d2001e1cc9ef7ff
# good: [1766c845a88dd9d4f303b8c1f639e9f2ac2aa4a2] Merge #318438: xkeyboard_config: 2.41 -> 2.42
git bisect good 1766c845a88dd9d4f303b8c1f639e9f2ac2aa4a2
# first bad commit: [737b5cd4f2f0113727312a12b4dfceee877a88c3] python3: 3.11.9 -> 3.12.4
```

### Notify maintainers

@mweinelt @Artturin @FRidh 

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/interpreters/python/cpython/3.12/0001-Fix-build-with-_PY_SHORT_FLOAT_REPR-0.patch
+++ b/pkgs/development/interpreters/python/cpython/3.12/0001-Fix-build-with-_PY_SHORT_FLOAT_REPR-0.patch
@@ -1,0 +1,53 @@
+From 04bfb877c8ccbd431dcae429abb487c1e3390801 Mon Sep 17 00:00:00 2001
+From: Yureka <yuka@yuka.dev>
+Date: Sun, 30 Jun 2024 09:37:49 +0200
+Subject: [PATCH] Fix build with _PY_SHORT_FLOAT_REPR == 0
+
+---
+ Include/internal/pycore_dtoa.h | 10 +++-------
+ 1 file changed, 3 insertions(+), 7 deletions(-)
+
+diff --git a/Include/internal/pycore_dtoa.h b/Include/internal/pycore_dtoa.h
+index 4d9681d59a..899d413b05 100644
+--- a/Include/internal/pycore_dtoa.h
++++ b/Include/internal/pycore_dtoa.h
+@@ -11,8 +11,6 @@ extern "C" {
+ #include "pycore_pymath.h"        // _PY_SHORT_FLOAT_REPR
+ 
+ 
+-#if _PY_SHORT_FLOAT_REPR == 1
+-
+ typedef uint32_t ULong;
+ 
+ struct
+@@ -22,15 +20,15 @@ Bigint {
+     ULong x[1];
+ };
+ 
+-#ifdef Py_USING_MEMORY_DEBUGGER
++#if defined(Py_USING_MEMORY_DEBUGGER) || _PY_SHORT_FLOAT_REPR == 0
+ 
+ struct _dtoa_state {
+     int _not_used;
+ };
+-#define _dtoa_interp_state_INIT(INTERP) \
++#define _dtoa_state_INIT(INTERP) \
+     {0}
+ 
+-#else  // !Py_USING_MEMORY_DEBUGGER
++#else  // !Py_USING_MEMORY_DEBUGGER && _PY_SHORT_FLOAT_REPR != 0
+ 
+ /* The size of the Bigint freelist */
+ #define Bigint_Kmax 7
+@@ -65,8 +63,6 @@ PyAPI_FUNC(char *) _Py_dg_dtoa(double d, int mode, int ndigits,
+                         int *decpt, int *sign, char **rve);
+ PyAPI_FUNC(void) _Py_dg_freedtoa(char *s);
+ 
+-#endif // _PY_SHORT_FLOAT_REPR == 1
+-
+ #ifdef __cplusplus
+ }
+ #endif
+-- 
+2.45.1
+

--- a/pkgs/development/interpreters/python/cpython/3.13/0001-Fix-build-with-_PY_SHORT_FLOAT_REPR-0.patch
+++ b/pkgs/development/interpreters/python/cpython/3.13/0001-Fix-build-with-_PY_SHORT_FLOAT_REPR-0.patch
@@ -1,0 +1,53 @@
+From 94d8a9efe6ec86a6e5b4806dbfb82ac926286456 Mon Sep 17 00:00:00 2001
+From: Yureka <yuka@yuka.dev>
+Date: Sun, 30 Jun 2024 09:45:58 +0200
+Subject: [PATCH] Fix build with _PY_SHORT_FLOAT_REPR == 0
+
+---
+ Include/internal/pycore_dtoa.h | 10 +++-------
+ 1 file changed, 3 insertions(+), 7 deletions(-)
+
+diff --git a/Include/internal/pycore_dtoa.h b/Include/internal/pycore_dtoa.h
+index c5cfdf4ce8..e4222c5267 100644
+--- a/Include/internal/pycore_dtoa.h
++++ b/Include/internal/pycore_dtoa.h
+@@ -11,8 +11,6 @@ extern "C" {
+ #include "pycore_pymath.h"        // _PY_SHORT_FLOAT_REPR
+ 
+ 
+-#if _PY_SHORT_FLOAT_REPR == 1
+-
+ typedef uint32_t ULong;
+ 
+ struct
+@@ -22,15 +20,15 @@ Bigint {
+     ULong x[1];
+ };
+ 
+-#ifdef Py_USING_MEMORY_DEBUGGER
++#if defined(Py_USING_MEMORY_DEBUGGER) || _PY_SHORT_FLOAT_REPR == 0
+ 
+ struct _dtoa_state {
+     int _not_used;
+ };
+-#define _dtoa_interp_state_INIT(INTERP) \
++#define _dtoa_state_INIT(INTERP) \
+     {0}
+ 
+-#else  // !Py_USING_MEMORY_DEBUGGER
++#else  // !Py_USING_MEMORY_DEBUGGER && _PY_SHORT_FLOAT_REPR != 0
+ 
+ /* The size of the Bigint freelist */
+ #define Bigint_Kmax 7
+@@ -66,8 +64,6 @@ extern char* _Py_dg_dtoa(double d, int mode, int ndigits,
+                          int *decpt, int *sign, char **rve);
+ extern void _Py_dg_freedtoa(char *s);
+ 
+-#endif // _PY_SHORT_FLOAT_REPR == 1
+-
+ 
+ extern PyStatus _PyDtoa_Init(PyInterpreterState *interp);
+ extern void _PyDtoa_Fini(PyInterpreterState *interp);
+-- 
+2.45.1
+

--- a/pkgs/development/interpreters/python/cpython/default.nix
+++ b/pkgs/development/interpreters/python/cpython/default.nix
@@ -367,7 +367,18 @@ in with passthru; stdenv.mkDerivation (finalAttrs: {
     };
   in [
     "${mingw-patch}/*.patch"
-  ]);
+  ]) ++ optionals (pythonAtLeast "3.12" && (stdenv.hostPlatform != stdenv.buildPlatform) && (
+    stdenv.hostPlatform.isAarch32 || stdenv.hostPlatform.isAarch64 || stdenv.hostPlatform.isRiscV
+  )) [
+    # backport fix for various platforms; armv7l, riscv64
+    # https://github.com/python/cpython/pull/121178
+    (
+      if (pythonAtLeast "3.13") then
+        ./3.13/0001-Fix-build-with-_PY_SHORT_FLOAT_REPR-0.patch
+      else
+        ./3.12/0001-Fix-build-with-_PY_SHORT_FLOAT_REPR-0.patch
+    )
+  ];
 
   postPatch = optionalString (!stdenv.hostPlatform.isWindows) ''
     substituteInPlace Lib/subprocess.py \

